Un-archive our opam scripts!

dra27 opened the pull request:
Part of this commit is being used for the test opam-repository for Relocatable OCaml at [dra27/opam-repository#relocatable](https://github.com/dra27/opam-repository/tree/relocatable/packages/ocaml), but it's not part of Relocatable itself. This is also related - and should be updated with - the equivalent scripts for `ocaml-system` (cf. ocaml/opam-repository#28455 and ocaml/opam-source-archives#52).

opam 2.0 generalised the packaging of the compiler and introduced the `ocaml` meta-package. This virtual package has a few variables associated with it which can be seen if you run `opam config list ocaml` in a switch:

```console
ocaml:native         true
ocaml:native-tools   true
ocaml:native-dynlink true
ocaml:stubsdir       <switch>\lib\ocaml\..\stublibs;<switch>\lib\ocaml\stublibs;<switch>\lib\ocaml
ocaml:preinstalled   false
ocaml:compiler       5.4.0
```

These variables are probed by an OCaml script run by the `ocaml` package itself. A copy of this script was stored with each of the `ocaml` packages versions in opam-repository which was a bit tedious when it needed updating (cf. https://github.com/ocaml/opam-repository/pull/11928) and so I aggregated it in https://github.com/ocaml/opam-repository/pull/12832 to create the `ocaml-config` package, which reduced the duplication. The `ocaml-config` existed _solely_ to contain this script in opam-repository.

Since last year, opam-repository (very sensibly) does not allow scripts and other files to be stored directly in the repository, for reasons that don't need listing here. The scripts were therefore moved as part of https://github.com/ocaml/opam-repository/pull/25960 to [ocaml/opam-source-archives](https://github.com/ocaml/opam-source-archives/tree/main/patches/ocaml-config). There are two things about this which I find unfortunate and which this PR seeks to fix:
1. A repository with word "archives" in it is in fact the primary source of a maintenance script which is executed by _every single installation of OCaml_. That's a bit strange...
2. The `ocaml-config` package was a bit of a hack necessary to stop duplicating the file in opam-repository - the switch to refer to it with a download URL actually removes the need for a separate `ocaml-config` package.

The first six commits in this branch simply cherry-pick the script and `ocaml-config.install` file from opam-repository, placing the result in `tools/opam`. Note that the same name is used throughout - the `ocaml-config` packages would simply refer to the appropriate commit in the `extra-source` section. For example, [https://raw.githubusercontent.com/ocaml/opam-source-archives/main/patches/ocaml-config/gen\_ocaml\_config.ml.in.3](https://raw.githubusercontent.com/ocaml/opam-source-archives/main/patches/ocaml-config/gen_ocaml_config.ml.in.3) (sha256 a9ad8d84a08961159653a978db92d10f694510182b206cacb96d5c9f63b5121e) can instead refer to [https://raw.githubusercontent.com/ocaml/ocaml/ec64d4913170fda621fe77bef77c596e5f50cb47/tools/opam/gen\_ocaml\_config.ml.in](https://raw.githubusercontent.com/ocaml/ocaml/ec64d4913170fda621fe77bef77c596e5f50cb47/tools/opam/gen\_ocaml\_config.ml.in).

The seventh commit ("Fully unify and update gen_ocaml_config.ml") refactors the script to eliminate the use of opam's substs mechanism. That requires changes to the `ocaml` packages, a demo of which is in https://github.com/dra27/opam-repository/commit/dc288c2320af5c5451fc831c59bba72f1da792ea (only the changes to the `ocaml` package in that commit matter), but it essentially removes the `ocaml-config` package completely and puts the variable expansions explicitly into the build recipe, so we have this in ocaml.5.4.0:

```
build: [
  ["ocaml" "gen_ocaml_config.ml" _:version _:name
     # The packages referred to in this string must be in the disjunction above
     # and also all be in the ocaml-core-compiler conflict-class
     "%{ocaml-system:installed?system}%%{ocaml-base-compiler:version}%%{dkml-base-compiler:version}%%{ocaml-variants:version}%"
     ocaml-system:installed
     # The order of these parameters matches the "convention" for the
     # ocaml-options-only packages
     "%{ocaml-option-32bit:installed?+32bit:}%"
     "%{ocaml-option-afl:installed?+afl:}%"
     "%{ocaml-option-bytecode-only:installed?+bytecode-only:}%"
     "%{ocaml-option-fp:installed?+fp:}%"
     "%{ocaml-option-flambda:installed?+flambda:}%"
     "%{ocaml-option-musl:installed?+musl:}%"
     "%{ocaml-option-no-flat-float-array:installed?+no-flat-float-array:}%"
     "%{ocaml-option-static:installed?+static:}%"]
]
```

where before we had this unfortunate layout in `gen_ocaml_config.ml.in` (note the doubling of `%` for actual `printf`s, and so forth:

```ocaml

  let p fmt = Printf.fprintf oc (fmt ^^ "\n") in
  p "opam-version: \"2.0\"";
  p "variables {";
  p "  native: %%b"
    (Sys.file_exists (ocaml^"opt"^suffix));
  p "  native-tools: %%b"
    (* The variable [ocamlc] already has a suffix on Windows
       (ex. '...\bin\ocamlc.exe') so we use [ocaml] to check *)
    (Sys.file_exists (ocaml^"c.opt"^suffix));
  p "  native-dynlink: %%b"
    has_native_dynlink;
  p "  stubsdir: %%S"
    stubsdir;
  p "  preinstalled: %{ocaml-system:installed}%";
  p "  compiler: \"%{ocaml-system:installed?system:}%%{ocaml-base-compiler:version}%%{dkml-base-compiler:version}%%{ocaml-variants:version}%%{ocaml-option-32bit:installed?+32bit:}%%{ocaml-option-afl:installed?+afl:}%%{ocaml-option-bytecode-only:installed?+bytecode-only:}%%{ocaml-option-default-unsafe-string:installed?+default-unsafe-string:}%%{ocaml-option-fp:installed?+fp:}%%{ocaml-option-flambda:installed?+flambda:}%%{ocaml-option-musl:installed?+musl:}%%{ocaml-option-nnp:installed?+nnp:}%%{ocaml-option-no-flat-float-array:installed?+no-flat-float-array:}%%{ocaml-option-spacetime:installed?+spacetime:}%%{ocaml-option-static:installed?+static:}%\"";
  p "}";
```

The last commit then updates the script to be able to cope with the relative paths proposed in #14243.

In terms of review, the first six commits are (immutable) history - the key thing is that the files in them are the same as those stored in opam-source-archives. The seventh commit is then best viewed as a "new" script that we (OCaml core devs) are "taking ownership" (I'd say blame is more appropriate!) and then the last commit is the proposed alteration for Relocatable OCaml.

- [x] Relocatable OCaml commit either confirmed or moved to a separate PR
- [x] Follow-up PR to opam-repository to switch the URLs from ocaml/opam-source-archives to ocaml/ocaml
- [x] `tools/opam/Layout.md`
- [x] Update with `ocaml-system`'s `gen_ocaml_config.ml.in`

dra27 committed:
Unify and update opam package scripts

There was technically a slightly tailored version of
gen_ocaml_config.ml.in sitting in the history to support the OCaml 3.07
package (which is the oldest compiler archived in opam-repository). The
revised version now supports all versions of the ocaml package from a
single script.

Code slightly updated with comments and a few very minor style changes.
Errors now display actual messages rather than using failwith.

dra27 committed:
Pass variables explicitly to opam scripts

The ocaml-config package originally stored gen_ocaml_config.ml.in in the
files/ subdirectory, but policy changes in 2025 meaning it is now
downloaded from this repository. That removes the need for a separate
ocaml-config package, as its only purpose was to stop the script being
stored multiple times in the same repository.

Both gen_ocaml_config.ml.in and gen_ocaml-system_config.ml.in are
updated to have the opam variables they use passed explicitly on the
command line by the file using them. For the ocaml package, this means
that irrelevant options (for example spacetime) can be omitted and it
makes the scripts themselves somewhat easier to read.

dra27 committed:
Handle case of missing ld.conf

At present, ld.conf is always written (even for --disable-shared, as it
still adds the Standard Library directory to ld.conf), but harden the
script against installations which quite reasonably choose not to
include it.

dra27 committed:
Support relative paths in gen_ocaml_config.ml

dra27 committed:
Quote the ocamlc command for ocamlc -where

gasche reviewed (approved):
Okay!

kit-ty-kate commented on tools/opam/gen_ocaml_config.ml:
```diff
@@ -0,0 +1,161 @@
+      else
+        Sys.ocaml_version)
+  in
+  if ocaml_version <> expected_ocaml_version then
+    (Printf.eprintf
+       "OCaml version mismatch: %s, expected %s"
+       ocaml_version expected_ocaml_version;
+     exit 1)
```
As mentioned in https://github.com/ocaml/opam-repository/pull/28455, i don't think we should keep this check

dra27 replied:
This is a different script from that one. The check here is ensuring that, say, the ocaml.5.4.0 package is always installed  with a compiler reporting version 5.4.0: here it's a sanity check that I think we should keep. I haven't forgotten about pulling ocaml-system's version of this script into this PR.

dra27 committed:
Change opam scripts to be plain .ml files

Since the substs mechanism is no longer required, these can now be
simple .ml files rather than .ml.in, which also means they can be run
outside of opam more easily.

dra27 committed:
Documentation for opam packaging

dra27 commented:
I need to run a few little tests on this locally, but the tasks for this PR are now complete. The main PR text is somewhat out-of-sync with the actual series, since the history commits were extracted and merged in #14351.

I've added a rudimentary Layout.md documenting some of the various packages we/I maintain in opam-repository for the compiler and how the scripts in `tools/opam` interact with them.

The renaming of the .ml.in scripts to .ml makes the overall "diff" look like it just adds new scripts, but the individual commits are clearer. The six commits:
1. Tweak a few bits of error handling and style in the two .config-generating scripts and, in particular, update gen_ocaml_config.ml to be useable by all the ocaml packages in opam-repository
2. Switch from having, say, `%{_:version}%` embedded in the .ml file to reading that string from a command-line argument instead. This has the benefit of allowing...
3. The .ml.in files to be renamed to .ml files, which means that the scripts are directly runnable outside of an opam package installation (they contain comments showing what running them does as well)
4. Adds a hardening to gen\_ocaml\_config.ml from an earlier version of bfdaf1719b3b0780f63e5a9548421dea1d2a618c. Technically-speaking, static builds of the compiler don't need to install ld.conf at all. Although we still do so at the moment, I've hardened the script not to crash if no ld.conf file is found in the compiler's library directory; it just sets an empty search path (as the runtime itself does)
5. The quoting of the ocamlc is hardened
6. Relative paths in ld.conf are correctly supported.

A version of this needs to be in to be used for the ocaml.5.5.0 package for the the 5.5.0~alpha1 release - assuming no objections, I was planning on merging this once I've completed testing it.
